active/XamlServiceProvider/XamlServiceProviderApi.md

```diff
@@ -0,0 +1,272 @@
+
+A developer defines a custom markup extension by subclassing the [MarkupExtension](https://docs.microsoft.com/uwp/api/Windows.UI.Xaml.Markup.MarkupExtension) class and overriding the ProvideValue method. This can then be used in Xaml markup: the markup extension can be created using the "{ }" markup expression syntax, and the return value from ProvideValue will be set as the property value.
+
+For example, given this markup extension:
+
+```cs
+public class CurrentDate : MarkupExtension
+{
+    public string Format { get; set; } = "";
+    protected override object ProvideValue()
+    {
+        return DateTime.Now.ToString(Format);
+    }
+}
+```
+
+you can do this in Xaml markup:
+
+```xml
+<TextBlock>
+    Today is:  
+    <Run Text="{local:CurrentDate Format=d}"/>
+</TextBlock>
+```
+
+On an en-US machine, this will produce something like:
+ 
+![MarkupExtension sample](MESample.png)
+
+# Examples
+
+## IProvideValueTarget
+
+The following example shows a custom markup extension whose provided value changes based on the specific property being targeted by the custom markup extension. In this case, the developer wants something that will automatically use an AcrylicBrush for the [Control.Background](https://docs.microsoft.com/uwp/api/Windows.UI.Xaml.Controls.Control.Background) or [Border.Background](https://docs.microsoft.com/uwp/api/Windows.UI.Xaml.Controls.Border.Background) properties, but uses a [SolidColorBrush](https://docs.microsoft.com/uwp/api/Windows.UI.Xaml.Media.SolidColorBrush) for all other properties. The ProvideValue implementation deterimines what value to return by using the IProvideValueTarget it gets from the IXamlServiceProvider parameter passed to ProvideValue.
+
+```cs
+public class BrushSelectorExtension : MarkupExtension
+{
+    public Color Color { get; set; }
+
+    protected override object ProvideValue(IXamlServiceProvider serviceProvider)
+    {
+        Brush brushToReturn = new SolidColorBrush() { Color = Color }; ;
+
+        var provideValueTarget = serviceProvider.GetService(typeof(IProvideValueTarget))
+            as IProvideValueTarget;
+        if (provideValueTarget.TargetProperty is ProvideValueTargetProperty targetProperty)
+        {
+            if (targetProperty.Name == "Background"
+                && (targetProperty.DeclaringType == typeof(Control) 
+                    || targetProperty.DeclaringType == typeof(Border)))
+            {
+                brushToReturn = new AcrylicBrush()
+                    { TintColor = Color, TintOpacity = 0.75 };
+            }
+        }
+
+        return brushToReturn;
+    }
+}
+```
+
+```xml
+<StackPanel>
+    <Button Foreground="{local:BrushSelector Color=Blue}" 
+            Background="{local:BrushSelector Color=Gold}">
+        Go bears!
+    </Button>
+    <Rectangle x:Name="SolidColor" Fill="{local:BrushSelector Color=Green}" />
+</StackPanel>
+```
+
+## IRootObjectProvider
+
+> Some more realistic example?

```

Comment:
How about using the RootObject as a default data source?

```
public class TestMarkupExtension : MarkupExtension
{
    string PropertyName { get; set; }
    object Source { get; set; }
    public override object ProvideValue(IXamlServiceProvider serviceProvider)
    {
        var source = Source ?? ((IRootObjectProvider)serviceProvider.GetService(typeof(IRootObjectProvider))).RootObject;
        var info = source.GetType().GetProperty(PropertyName);
        return info.GetValue(target);
    }
}
```

Reply:
Actually I'm not understanding this. If getting a PropertyInfo from Source, and then using it to read from 'target'. I'm not sure if that's IPropertyValueTarget, or was meant to be 'source'. It kind of looks like it's manually doing an x:Bind.

Reply:
`Direction="{local:MyBind PropertyName=DirectionToUse, Source=Somebody}"` will set the Direction property to the current value of `Somebody.DirectionToUser`. And if you don't specify a Source, then it defaults to the root page. So yeah, this is basically x:Bind reimplemented, but at least it's something less artificial than "Give me the title of the root element."

Reply:
Cool, basically a dynamic x:Bind